wagtail/project_template/Dockerfile

```diff
@@ -1,27 +1,50 @@
-# Use an official Python runtime as a parent image
-FROM python:3.7
-LABEL maintainer="hello@wagtail.io"
+# Use an official Python runtime based on Debian 10 "buster" as a parent image.
+FROM python:3.8.1-buster
```

Comment:
Should we use -slim instead? See https://pythonspeed.com/articles/alpine-docker-python/ and https://news.ycombinator.com/item?id=22182226

Reply:
I can verify later if it includes `libpq-dev` and  Pillow dependencies. I assume we want to keep those in, right?

Reply:
Yes, we do, good point.

Reply:
For the Pillow dependencies:
```
root@c9f81c52d774:/app# apt-get install libtiff5-dev libjpeg62-turbo-dev libopenjp2-7-dev zlib1g-dev     libfreetype6-dev liblcms2-dev libwebp-dev tcl8.6-dev  
tk8.6-dev python3-tk     libharfbuzz-dev libfribidi-dev                                                                                                        
Reading package lists... Done                                                  
Building dependency tree                                                       
Reading state information... Done                                              
The following additional packages will be installed:
  binutils binutils-common binutils-x86-64-linux-gnu blt build-essential bzip2 cpp cpp-8 dirmngr dpkg-dev fakeroot file fontconfig-config fonts-dejavu-core    
  g++ g++-8 gcc gcc-8 gir1.2-glib-2.0 gir1.2-harfbuzz-0.0 gnupg gnupg-l10n gnupg-utils gpg gpg-agent gpg-wks-client gpg-wks-server gpgconf gpgsm                                                                                                                                                                              
  icu-devtools libalgorithm-diff-perl libalgorithm-diff-xs-perl libalgorithm-merge-perl libasan5 libassuan0 libatomic1 libbinutils libblkid-dev libbsd0                                                                                                                                                                       
  libc-dev-bin libc6-dev libcc1-0 libdpkg-perl libdrm-amdgpu1 libdrm-common libdrm-intel1 libdrm-nouveau2 libdrm-radeon1 libdrm2 libedit2 libelf1                                                                                                                                                                             
  libexpat1-dev libfakeroot libffi-dev libfile-fcntllock-perl libfontconfig1 libfontconfig1-dev libfontenc1 libfreetype6 libfribidi0 libgcc-8-dev                                                                                                                                                                             
  libgdbm-compat4 libgirepository-1.0-1 libgl1 libgl1-mesa-dri libglapi-mesa libglib2.0-0 libglib2.0-bin libglib2.0-data libglib2.0-dev libglib2.0-dev-bin                                                                                                                                                                    
  libglvnd0 libglx-mesa0 libglx0 libgomp1 libgraphite2-3 libgraphite2-dev libharfbuzz-gobject0 libharfbuzz-icu0 libharfbuzz0b libice-dev libice6 libicu-dev                                                                                                                                                                   
  libicu63 libisl19 libitm1 libjbig-dev libjbig0 libjpeg62-turbo libksba8 liblcms2-2 libldap-2.4-2 libldap-common libllvm7 liblocale-gettext-perl liblsan0                                                                                                                                                                    
  liblzma-dev libmagic-mgc libmagic1 libmount-dev libmpc3 libmpdec2 libmpfr6 libmpx2 libnpth0 libopenjp2-7 libpciaccess0 libpcre16-3 libpcre3-dev                                                                                                                                                                             
  libpcre32-3 libpcrecpp0v5 libperl5.28 libpng-dev libpng-tools libpng16-16 libpthread-stubs0-dev libpython3-stdlib libpython3.7-minimal libpython3.7-stdlib                                                                                                                                                                  
  libquadmath0 libsasl2-2 libsasl2-modules libsasl2-modules-db libselinux1-dev libsensors-config libsensors5 libsepol1-dev libsm-dev libsm6 libstdc++-8-dev                                                                                                                                                                   
  libtcl8.6 libtiff-dev libtiff5 libtiffxx5 libtk8.6 libtsan0 libubsan1 libutempter0 libwebp6 libwebpdemux2 libwebpmux3 libx11-6 libx11-data libx11-dev                                                                                                                                                                       
  libx11-xcb1 libxau-dev libxau6 libxaw7 libxcb-dri2-0 libxcb-dri3-0 libxcb-glx0 libxcb-present0 libxcb-shape0 libxcb-sync1 libxcb1 libxcb1-dev                                                                                                                                                                               
  libxcomposite1 libxdamage1 libxdmcp-dev libxdmcp6 libxext-dev libxext6 libxfixes3 libxft-dev libxft2 libxi6 libxinerama1 libxml2 libxmu6 libxmuu1 libxpm4                                                                                                                                                                   
  libxrandr2 libxrender-dev libxrender1 libxshmfence1 libxss-dev libxss1 libxt-dev libxt6 libxtst6 libxv1 libxxf86dga1 libxxf86vm1 linux-libc-dev lsb-base                                                                                                                                                                    
  make manpages manpages-dev mime-support patch perl perl-modules-5.28 pinentry-curses pkg-config python3 python3-distutils python3-lib2to3 python3-minimal                                                                                                                                                                   
  python3.7 python3.7-minimal sensible-utils shared-mime-info tcl8.6 tk8.6 tk8.6-blt2.5 ucf uuid-dev x11-common x11-utils x11proto-core-dev x11proto-dev                                                                                                                                                                      
  x11proto-scrnsaver-dev x11proto-xext-dev xbitmaps xdg-user-dirs xorg-sgml-doctools xterm xtrans-dev xz-utils                                                                                                                                                                                                                
Suggested packages:                                                                                                                                                                                                                                                                                                           
  binutils-doc blt-demo bzip2-doc cpp-doc gcc-8-locales dbus-user-session libpam-systemd pinentry-gnome3 tor debian-keyring g++-multilib g++-8-multilib                                                                                                                                                                       
  gcc-8-doc libstdc++6-8-dbg gcc-multilib autoconf automake libtool flex bison gdb gcc-doc gcc-8-multilib libgcc1-dbg libgomp1-dbg libitm1-dbg                                                                                                                                                                                
  libatomic1-dbg libasan5-dbg liblsan0-dbg libtsan0-dbg libubsan1-dbg libmpx2-dbg libquadmath0-dbg parcimonie xloadimage scdaemon glibc-doc git bzr                                                                                                                                                                           
  freetype2-doc libglib2.0-doc libgdk-pixbuf2.0-bin | libgdk-pixbuf2.0-dev libxml2-utils libgraphite2-utils libice-doc icu-doc liblcms2-utils liblzma-doc                                                                                                                                                                     
  pciutils libsasl2-modules-gssapi-mit | libsasl2-modules-gssapi-heimdal libsasl2-modules-ldap libsasl2-modules-otp libsasl2-modules-sql lm-sensors                                                                                                                                                                           
  libsm-doc libstdc++-8-doc libx11-doc libxcb-doc libxext-doc libxt-doc make-doc man-browser ed diffutils-doc perl-doc libterm-readline-gnu-perl                                                                                                                                                                              
  | libterm-readline-perl-perl libb-debug-perl liblocale-codes-perl pinentry-doc python3-doc python3-venv tix python3-tk-dbg python3.7-venv python3.7-doc                                                                                                                                                                     
  binfmt-support tcl-tclreadline tcl8.6-doc tk8.6-doc mesa-utils xfonts-cyrillic                                                                                                                                                                                                                                              
The following NEW packages will be installed:                                                                                                                                                                                                                                                                                 
  binutils binutils-common binutils-x86-64-linux-gnu blt build-essential bzip2 cpp cpp-8 dirmngr dpkg-dev fakeroot file fontconfig-config fonts-dejavu-core                                                                                                                                                                   
  g++ g++-8 gcc gcc-8 gir1.2-glib-2.0 gir1.2-harfbuzz-0.0 gnupg gnupg-l10n gnupg-utils gpg gpg-agent gpg-wks-client gpg-wks-server gpgconf gpgsm                                                                                                                                                                              
  icu-devtools libalgorithm-diff-perl libalgorithm-diff-xs-perl libalgorithm-merge-perl libasan5 libassuan0 libatomic1 libbinutils libblkid-dev libbsd0                                                                                                                                                                       
  libc-dev-bin libc6-dev libcc1-0 libdpkg-perl libdrm-amdgpu1 libdrm-common libdrm-intel1 libdrm-nouveau2 libdrm-radeon1 libdrm2 libedit2 libelf1                                                                                                                                                                             
  libexpat1-dev libfakeroot libffi-dev libfile-fcntllock-perl libfontconfig1 libfontconfig1-dev libfontenc1 libfreetype6 libfreetype6-dev libfribidi-dev                                                                                                                                                                      
  libfribidi0 libgcc-8-dev libgdbm-compat4 libgirepository-1.0-1 libgl1 libgl1-mesa-dri libglapi-mesa libglib2.0-0 libglib2.0-bin libglib2.0-data                                                                                                                                                                             
  libglib2.0-dev libglib2.0-dev-bin libglvnd0 libglx-mesa0 libglx0 libgomp1 libgraphite2-3 libgraphite2-dev libharfbuzz-dev libharfbuzz-gobject0                                                                                                                                                                              
  libharfbuzz-icu0 libharfbuzz0b libice-dev libice6 libicu-dev libicu63 libisl19 libitm1 libjbig-dev libjbig0 libjpeg62-turbo libjpeg62-turbo-dev libksba8                                                                                                                                                                    
  liblcms2-2 liblcms2-dev libldap-2.4-2 libldap-common libllvm7 liblocale-gettext-perl liblsan0 liblzma-dev libmagic-mgc libmagic1 libmount-dev libmpc3                                                                                                                                                                       
  libmpdec2 libmpfr6 libmpx2 libnpth0 libopenjp2-7 libopenjp2-7-dev libpciaccess0 libpcre16-3 libpcre3-dev libpcre32-3 libpcrecpp0v5 libperl5.28 libpng-dev                                                                                                                                                                   
  libpng-tools libpng16-16 libpthread-stubs0-dev libpython3-stdlib libpython3.7-minimal libpython3.7-stdlib libquadmath0 libsasl2-2 libsasl2-modules                                                                                                                                                                          
  libsasl2-modules-db libselinux1-dev libsensors-config libsensors5 libsepol1-dev libsm-dev libsm6 libstdc++-8-dev libtcl8.6 libtiff-dev libtiff5                                                                                                                                                                             
  libtiff5-dev libtiffxx5 libtk8.6 libtsan0 libubsan1 libutempter0 libwebp-dev libwebp6 libwebpdemux2 libwebpmux3 libx11-6 libx11-data libx11-dev                                                                                                                                                                             
  libx11-xcb1 libxau-dev libxau6 libxaw7 libxcb-dri2-0 libxcb-dri3-0 libxcb-glx0 libxcb-present0 libxcb-shape0 libxcb-sync1 libxcb1 libxcb1-dev                                                                                                                                                                               
  libxcomposite1 libxdamage1 libxdmcp-dev libxdmcp6 libxext-dev libxext6 libxfixes3 libxft-dev libxft2 libxi6 libxinerama1 libxml2 libxmu6 libxmuu1 libxpm4                                                                                                                                                                   
  libxrandr2 libxrender-dev libxrender1 libxshmfence1 libxss-dev libxss1 libxt-dev libxt6 libxtst6 libxv1 libxxf86dga1 libxxf86vm1 linux-libc-dev lsb-base                                                                                                                                                                    
  make manpages manpages-dev mime-support patch perl perl-modules-5.28 pinentry-curses pkg-config python3 python3-distutils python3-lib2to3 python3-minimal                                                                                                                                                                   
  python3-tk python3.7 python3.7-minimal sensible-utils shared-mime-info tcl8.6 tcl8.6-dev tk8.6 tk8.6-blt2.5 tk8.6-dev ucf uuid-dev x11-common x11-utils                                                                                                                                                                     
  x11proto-core-dev x11proto-dev x11proto-scrnsaver-dev x11proto-xext-dev xbitmaps xdg-user-dirs xorg-sgml-doctools xterm xtrans-dev xz-utils zlib1g-dev                                                                                                                                                                      
0 upgraded, 231 newly installed, 0 to remove and 0 not upgraded.                                                                                                                                                                                                                                                              
Need to get 151 MB of archives.                                                                                                                                                                                                                                                                                               
After this operation, 686 MB of additional disk space will be used.                                                                                                                                                                                                                                                           
Do you want to continue? [Y/n] 
```
We may not need all the Pillow dependencies but I don't think there's a value in spending too much time trying to figure it out. I think we should stick to the normal buster image. 

`libpq-dev` may not be mandatory in our use case, but we can expect that probably most Wagtail projects will use PostgreSQL, so that's a bugger.

Reply:
> `apt-get install libtiff5-dev libjpeg62-turbo-dev libopenjp2-7-dev zlib1g-dev     libfreetype6-dev liblcms2-dev libwebp-dev tcl8.6-dev  
tk8.6-dev python3-tk     libharfbuzz-dev libfribidi-dev                                                                                                        `

Is that the full set of dependencies from the Pillow docs? That's definitely overkill - the relevant ones for Wagtail are libjpeg, zlib1g-dev and libwebp-dev.

Reply:
Thanks @gasman and @tomdyson. I am able to upload all supported image types (without animated GIFs, do we want to support that out of the box?). I've switched to slim now. I am unsure if we want to install dependencies for MySQL, PostgreSQL and `build-essential` so they can install `psycopg` and MySQL out of the box

Reply:
I believe the non-slim python docker images contain a good amount of build deps for installing `mysqlclient`, `psycopg2`, etc. It would probably be "safer" to use the non-slim as it will support most developer's environments. It is pretty huge though so there is that downside, but it might make a better default.

Reply:
Yea, I personally am not bothered which way we go with (slim or non-slim). With the slim image we must be mindful not to make experience worse by developers having to research what system libraries they are missing. That's why I added some essential dependencies for database support, are people happy with it?

Reply:
It looks like installing the build-essentials package is actually pulling in Debian's packaged version of python (3.7) as a dependency. So I would be mindful about what is actually happening here as you are now bundling two competing versions of python (3.8 from the base image, and 3.7 from your RUN command). I think the non-slim base image would handle this better. See the list of packages getting installed in comment above https://github.com/wagtail/wagtail/pull/5800#discussion_r372608796

Reply:
I can't replicate apt-get pulling an extra Python version with the current Dockerfile. I am not installing all the packages above, only the specific Pillow dependencies pointed out by Matthew.
https://gist.github.com/tm-kn/4984d926fe44f97026eeb5e863c7d73d